src/useAddonState/manager.ts

```diff
@@ -0,0 +1,54 @@
+import type { API } from "@storybook/manager-api";
+import { useStorybookApi } from "@storybook/manager-api";
+import { useCallback } from "react";
+
+import { getValue, GetValuePayload, setValue, SetValuePayload } from "./common";
+
+const requestedKeys = new Set<string>();
+let listening = false;
+function ensureListening(api: API) {
+  if (listening) return;
+
+  api.on(getValue, ({ key }: GetValuePayload) => {
+    const apiKey = `experimental_${key}`;
+    const value = api.getAddonState(apiKey);
+    if (value !== undefined) api.emit(setValue, { key, value } satisfies SetValuePayload);
+  });
+
+  api.on(setValue, ({ key, value }: SetValuePayload) => {
+    const apiKey = `experimental_${key}`;
+    api.setAddonState(apiKey, value);
+  });
+
+  listening = true;
+}
+
+function apiGetAddonState<T>(api: API, key: string) {
+  const apiKey = `experimental_${key}`;
+
+  // Ask the universe for the value of the key if we've never seen it before
+  if (!requestedKeys.has(key)) {
+    api.emit(getValue, { key } satisfies GetValuePayload);
+    requestedKeys.add(key);
+  }
+
+  return api.getAddonState(apiKey) as T | undefined;
+}
+function apiSetAddonState<T>(api: API, key: string, value: T) {
+  const apiKey = `experimental_${key}`;
+  api.setAddonState(apiKey, value);
+
+  api.emit(setValue, { key, value } satisfies SetValuePayload);
+}
+
+export function useAddonState<T>(key: string) {
+  const api = useStorybookApi();
+
+  // This should probably just always be happening
```

Comment:
What does this comment mean?

Reply:
I think probably in SB we want to setup one copy of these listeners when the manager boots up and just always write values to the store/respond to requests for values, rather than only do it "on demand".